### PR TITLE
add new compiler options

### LINKED ISSUE #1800: Document the C# compiler /pathmap command line flag
Replacing https://github.com/dotnet/roslyn/issues/18071

Please see that issue for details on the new flag.

The new documentation should be under this directory https://github.com/dotnet/docs/tree/master/docs/csharp/language-reference/compiler-options

/cc @gafter

### LINKED ISSUE #3242: New compiler option: Strong name provider
More details coming. This has not been merged to master yet.

### LINKED ISSUE #4625: Add documentation for -publicsign compiler switch
# Issue Title

The csc switch `-publicsign` is missing documentation.

# General

This switch is used to apply a public key without actually signing it, like `-delaysign` but unlike `-delaysign`, `-publicsign` sets a bit in the assembly file that tells the runtime that the file is actually signed.  
This switch was added with https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/compiler-options/listed-by-category.md.

## Issues with Existing Topics

We should add the switch here: https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/compiler-options/listed-by-category.md

## Requests for new Topics

1. Tell us where this topic should go in the Table of Contents.

   - Follow the same pattern as is used https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/compiler-options/delaysign-compiler-option.md.

2. Write an abstract

   - Cribbed from https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/compiler-options/delaysign-compiler-option.md

   > Sometimes called "fake sign" or "OSS sign" public signing is including
the public key in an output assembly and setting the "signed" flag, but
not actually signing the assembly with a private key. This is useful for
open source projects where people want to build assemblies which are
compatible with the released "fully signed" assemblies, but don't have
access to the private key used to sign the assemblies. Since almost no
consumers actually need to check if the assembly is fully signed, these
publicly built assemblies are useable in almost every scenario that the
fully signed one would be used in.

3. Fill in an outline

    - Copy `-delaysign` outline

4. Suggest reviewers

    - @agocke @nguerrera @DnlHarvey

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #1800 
Fixes #3242 
Fixes #4625 

The new options are:

-pathmap
-publicsign
-keyfile now works on CoreCLR (-keycontainer does not)

Update lists of alphabetic commands, pages by category, and TOC.

Also, fixed a variety of markdown lint warnings. These are mostly whitespace.

Reviews will be easier using the rich diff view.

/cc @jcouv

### CODE REVIEW

**Comment by jcouv** [comment]
Tagging @tmat @agocke for review.

**Review by agocke — APPROVED** [review]

File: docs/csharp/language-reference/compiler-options/pathmap-compiler-option.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+f1_keywords: 
+  - "/pathmap"
+helpviewer_keywords: 
+  - "-pathmap compiler option [C#]"
+  - "pathmap compiler option [C#]"
+  - "/pathmap compiler option [C#]"
+---
+# -pathmap (C# Compiler Options)
+
+The **-pathmap** compiler option mapping for source path names output by the compiler.
```
Comment (svick) [inline docs/csharp/language-reference/compiler-options/pathmap-compiler-option.md]: Is this missing a verb (possibly "specifies")?

File: docs/csharp/language-reference/compiler-options/publicsign-compiler-option.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+f1_keywords: 
+  - "/publicsign"
+helpviewer_keywords: 
+  - "-publicsign compiler option [C#]"
+  - "publicsign compiler option [C#]"
+  - "/publicsign compiler option [C#]"
+---
+# -publicsign (C# Compiler Options)
+
+This option causes the compiler to apply a public key without actually signing it. The **-publicsign** option also sets a bit in the assembly that tells the runtime that the file is actually signed.
```
Comment (svick) [inline docs/csharp/language-reference/compiler-options/publicsign-compiler-option.md]: In the first sentence, I think it's not clear what "it" refers to.

File: docs/csharp/language-reference/compiler-options/publicsign-compiler-option.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+None.
+
+## Remarks
+
+The **-publicsign** option requires the use of the [-keyfile](keyfile-compiler-option.md) or [-keycontainer](keycontainer-compiler-option.md). The **keyfile** or **keycontainer** options specify the public key.
+
+The **-publicsign** and **-delaysign** options are mutually exclusive.
+
+Sometimes called "fake sign" or "OSS sign" public signing is including the public key in an output assembly and setting the "signed" flag, but not actually signing the assembly with a private key. This is useful for open source projects where people want to build assemblies which are compatible with the released "fully signed" assemblies, but don't have access to the private key used to sign the assemblies. Since almost no consumers actually need to check if the assembly is fully signed, these publicly built assemblies are useable in almost every scenario that the fully signed one would be used in.
```
Comment (svick) [inline docs/csharp/language-reference/compiler-options/publicsign-compiler-option.md]: Missing comma before "public signing" in the first sentence?

File: docs/csharp/language-reference/compiler-options/publicsign-compiler-option.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+f1_keywords: 
+  - "/publicsign"
+helpviewer_keywords: 
+  - "-publicsign compiler option [C#]"
+  - "publicsign compiler option [C#]"
+  - "/publicsign compiler option [C#]"
+---
+# -publicsign (C# Compiler Options)
+
+This option causes the compiler to apply a public key without actually signing the assembly. The **-publicsign** option also sets a bit in the assembly that tells the runtime that the file is actually signed.
```
Comment (JRAlexander) [inline docs/csharp/language-reference/compiler-options/publicsign-compiler-option.md]: Ing. Maybe "This option causes the compiler to apply a public key without an assembly sign."?

**Review by JRAlexander — CHANGES_REQUESTED** [review]
Looks good, @BillWagner - just a few small things.

File: docs/csharp/language-reference/compiler-options/keycontainer-compiler-option.md [hunk 5]
```diff
@@ -23,7 +23,7 @@ Specifies the name of the cryptographic key container.
  The name of the strong name key container.  
   
 ## Remarks  
- When the **-keycontainer** option is used, the compiler creates a sharable component by inserting a public key from the specified container into the assembly manifest and signing the final assembly with the private key. To generate a key file, type sn -k `file` at the command line. sn -i installs the key pair into a container.  
+ When the **-keycontainer** option is used, the compiler creates a sharable component by inserting a public key from the specified container into the assembly manifest and signing the final assembly with the private key. To generate a key file, type sn -k `file` at the command line. sn -i installs the key pair into a container. This option is not supported when the compiler is running on CoreCLR. To sign an assembly when building on CoreCLR, use the [-keyfile](keyfile-compiler-option.md) option.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/compiler-options/keycontainer-compiler-option.md]: I changed it a bit differently, but all the -ings are gone.

**Review by JRAlexander — APPROVED** [review]

**Comment by JRAlexander** [comment]
Excellent, @BillWagner!

File: docs/csharp/language-reference/compiler-options/listed-alphabetically.md [hunk 6]
```diff
@@ -9,69 +9,73 @@ helpviewer_keywords:
...
+|[@](response-file-compiler-option.md)|Reads a response file for more options.|
+|[-?](help-compiler-option.md)|Displays a usage message to stdout.|
+|-additionalfile|Names additional files that don't directly affect code generation but may be used by analyzers for producing errors or warnings.|
+|[-addmodule](addmodule-compiler-option.md)|Links the specified modules into this assembly|
+|-analyzer|Run the analyzers from this assembly (Short form: -a)|
+|[-appconfig](appconfig-compiler-option.md)|Specifies the location of app.config at assembly binding time.|
+|[-baseaddress](baseaddress-compiler-option.md)|Specifies the base address for the library to be built.|
+|[-bugreport](bugreport-compiler-option.md)|Creates a 'Bug Report' file. This file will be sent together with any crash information if it is used with -errorreport:prompt or -errorreport:send.|
+|[-checked](checked-compiler-option.md)|Causes the compiler to generate overflow checks.|
+|-checksumalgorithm:\<alg>|Specify the algorithm for calculating the source file checksum stored in PDB.  Supported values are: SHA1 (default) or SHA256.|
```
Comment (mairaw) [inline docs/csharp/language-reference/compiler-options/listed-alphabetically.md]: all the others are in 3rd person, why change here?

File: docs/csharp/language-reference/compiler-options/pathmap-compiler-option.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+## Example
+
+Compile `t.cs` in the directory **C:\\work\\tests** and map that directory to **\publish** in the output:
+
+```console
+csc -pathmap:C:\work\tests=\publish t.cs
+```
+
+## See Also
```
Comment (mairaw) [inline docs/csharp/language-reference/compiler-options/pathmap-compiler-option.md]: nit: sentence case

File: docs/csharp/language-reference/compiler-options/listed-alphabetically.md (hunk 6 above)
Comment (BillWagner) [reply docs/csharp/language-reference/compiler-options/listed-alphabetically.md]: I'll address this in a new PR.

I'm not seeing a change in person here. What am I missing?

Comment (mairaw) [reply docs/csharp/language-reference/compiler-options/listed-alphabetically.md]: It used to say specifies and it was matching the other lines - Causes, Links, Names, Displays, etc. 

Makes sense @BillWagner?

**Pull request merged**